File: backend/api/identity_admin.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.orm import Session

from auth.dependencies import require_permissions
from auth.identity_admin_service import identity_admin_service
from auth.rbac import Permission
from auth.schemas import Principal
from database import get_db


router = APIRouter(prefix="/api/admin/identity", tags=["identity-administration"])


class ProviderUpsertRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider_type: str
    display_name: str = Field(min_length=1, max_length=120)
    enabled: bool = False
    config: dict[str, Any] = Field(default_factory=dict)
    secrets: dict[str, str] = Field(default_factory=dict)
    clear_secrets: set[str] = Field(default_factory=set)
    group_role_mapping: dict[str, Any] = Field(default_factory=dict)


class UserUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    active: bool
    display_name: str = Field(min_length=1, max_length=160)
    email: str | None = Field(default=None, max_length=320)
    manual_roles: set[str] = Field(default_factory=set)


@router.get("/providers")
def list_identity_providers(
    principal: Principal = Depends(require_permissions(Permission.IDENTITIES_MANAGE.value)),
    db: Session = Depends(get_db),
):
    del principal
    return {"items": identity_admin_service.list_providers(db)}


@router.put("/providers/{provider_key}")
def upsert_identity_provider(
    provider_key: str,
    payload: ProviderUpsertRequest,
    principal: Principal = Depends(require_permissions(Permission.IDENTITIES_MANAGE.value)),
    db: Session = Depends(get_db),
):
    try:
        return identity_admin_service.upsert_provider(
            db,
            provider_key=provider_key,
            provider_type=payload.provider_type,
            display_name=payload.display_name,
            enabled=payload.enabled,
            config=payload.config,
            secrets=payload.secrets,
            clear_secrets=payload.clear_secrets,
            group_role_mapping=payload.group_role_mapping,
            actor=principal,
        )
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/users")
def list_users(
    principal: Principal = Depends(require_permissions(Permission.USERS_MANAGE.value)),
    db: Session = Depends(get_db),
):
    del principal
    return {"items": identity_admin_service.list_users(db)}


@router.put("/users/{user_id}")
def update_user(
    user_id: int,
    payload: UserUpdateRequest,
    principal: Principal = Depends(require_permissions(Permission.USERS_MANAGE.value)),
    db: Session = Depends(get_db),
):
    try:
        return identity_admin_service.update_user(
            db,
            user_id=user_id,
            active=payload.active,
            display_name=payload.display_name,
            email=payload.email,
            manual_roles=payload.manual_roles,
            actor=principal,
        )
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

File: frontend/src/api/identityAdmin.ts
import axios from '@/api/http'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || '/api'
const base = API_BASE_URL.startsWith('http') ? `${API_BASE_URL}/api/admin/identity` : `${API_BASE_URL}/admin/identity`

export interface IdentityProviderView {
  id: number
  provider_key: string
  provider_type: string
  display_name: string
  enabled: boolean
  config: Record<string, unknown>
  secret_status: Record<string, boolean>
  group_role_mapping: Record<string, string[]>
}

export interface UserView {
  id: number
  username: string
  display_name: string
  email?: string
  active: boolean
  is_emergency: boolean
  roles: Array<{ role: string; source: string; provider_id?: number }>
}

export const identityAdminApi = {
  async providers() {
    return (await axios.get<{ items: IdentityProviderView[] }>(`${base}/providers`)).data.items
  },
  async saveProvider(providerKey: string, payload: Record<string, unknown>) {
    return (await axios.put<IdentityProviderView>(`${base}/providers/${providerKey}`, payload)).data
  },
  async users() {
    return (await axios.get<{ items: UserView[] }>(`${base}/users`)).data.items
  },
  async saveUser(userId: number, payload: Record<string, unknown>) {
    return (await axios.put<UserView>(`${base}/users/${userId}`, payload)).data
  }
}
